### PR TITLE
add autoscaled setting to roles to disable replica updates

### PR DESCRIPTION
**Pull request by grosser**
@craig-day @catkins 
@dragonfax @jonmoter 

screen shot 2017-12-26 at 10 57 36 am

### CODE REVIEW

**Comment by jonmoter** [comment]
So in this implementation, I see it's a property of the DeployGroupRole. Which means that we could use autoscaling for some deploy groups, but not others?

What's the behavior if someone clicks the "autoscaling" box but does not implement an autoscaler?

Another way we could implement it is if there's a `HorizontalPodAutoscaler` object in the role file, then Samson automatically treats that role as using autoscaling.

File: plugins/kubernetes/app/models/kubernetes/template_filler.rb [hunk 1]
```diff
@@ -31,8 +31,7 @@ def to_hash
             set_history_limit
           end
 
-          set_replica_target unless ['DaemonSet', 'Pod'].include?(kind)
-
+          set_replica_target unless ['DaemonSet', 'Pod'].include?(template[:kind])
```
Comment (craig-day) [inline plugins/kubernetes/app/models/kubernetes/template_filler.rb]: shouldn't this be checking the autoscale thing and not do it if the role is autoscaled?

Comment (grosser) [reply plugins/kubernetes/app/models/kubernetes/template_filler.rb]: the "set replicas" logic is in resource.rb where the actual update is being done, the template does not know if it is being used for a create or update and I wanted the create to still have the user-defined number of replicas

**Comment by grosser** [comment]
If a user sets autoscale but does not implement one then replicase will never be changed.

Having the autoscaled on the deploy-group-role is nice for a smooth rollout, but could lead to inconsistent state ... so how about we move it to the role ? ... so it's consistent ...

The autoscaler could be external, so I don't want make it required to be in the role file.

**Comment by grosser** [comment]
updated to be configured from role ... looks good ?

File: plugins/kubernetes/app/models/kubernetes/resource.rb [hunk 2]
```diff
@@ -96,9 +97,12 @@ def update
         expire_cache
       end
 
-      # hook for subclasses to override
       def template_for_update
-        @template
+        copy = @template.deep_dup
+        if @autoscaled && spec = copy[:spec]
+          spec.delete(:replicas)
```
Comment (craig-day) [inline plugins/kubernetes/app/models/kubernetes/resource.rb]: should this be `copy`?

File: plugins/kubernetes/app/models/kubernetes/resource.rb [hunk 3]
```diff
@@ -7,9 +7,10 @@ module Kubernetes
   # and see what it does internally ... simple create/update/delete requests or special magic ?
   module Resource
     class Base
-      def initialize(template, deploy_group)
+      def initialize(template, deploy_group, autoscaled:)
```
Comment (craig-day) [inline plugins/kubernetes/app/models/kubernetes/resource.rb]: any backwards compatibility concerns? Should it default to false?

File: plugins/kubernetes/test/models/kubernetes/resource_test.rb [hunk 4]
```diff
@@ -84,6 +85,17 @@ def assert_pod_deletion
       assert_requested get, times: 2
     end
 
+    it "does not set replicase when autoscaled, to not revert autoscaler changes" do
+      resource = Kubernetes::Resource.build(template, deploy_group, autoscaled: true)
+      get = stub_request(:get, url).to_return(body: '{}')
+      update = stub_request(:put, url).with {  |x| x.body.wont_include 'replica'; true }.to_return(body: "{}")
```
Comment (craig-day) [inline plugins/kubernetes/test/models/kubernetes/resource_test.rb]: extra space before the loop argument

File: plugins/kubernetes/test/models/kubernetes/resource_test.rb [hunk 5]
```diff
@@ -74,7 +75,7 @@ def assert_pod_deletion
     it "updates existing" do
       get = stub_request(:get, url).to_return(body: '{}')
 
-      update = stub_request(:put, url).to_return(body: "{}")
+      update = stub_request(:put, url).with {  |x| x.body.must_include 'replica'; true }.to_return(body: "{}")
```
Comment (craig-day) [inline plugins/kubernetes/test/models/kubernetes/resource_test.rb]: extra space here too

File: plugins/kubernetes/app/models/kubernetes/resource.rb (hunk 3 above)
Comment (grosser) [reply plugins/kubernetes/app/models/kubernetes/resource.rb]: nope, I want to force every callsite to know about this

File: plugins/kubernetes/app/models/kubernetes/resource.rb (hunk 2 above)
Comment (grosser) [reply plugins/kubernetes/app/models/kubernetes/resource.rb]: no, it's spec.replicas

Comment (craig-day) [reply plugins/kubernetes/app/models/kubernetes/resource.rb]: oh, I thought this read `spec == copy[:spec]`, but nvm...

**Comment by grosser** [comment]
how about now ?

**Comment by grosser** [comment]
... looks like not setting replicas is setting it to 1 :/

**Comment by grosser** [comment]
added an icon for it on the stage view so it's less confusing ... 
screen shot 2017-12-27 at 3 14 06 pm

**Pull request merged**